web/gridt/resources/user.py

```diff
@@ -45,4 +53,71 @@ def post(self):
         
         if current_identity.verify_password(res["old_password"]):
             current_identity.hash_password(res["new_password"])
-            return ({"message": "Successfully changed password."}, 200)
\ No newline at end of file
+            
+            link = "https://app.gridt.org/request_password_reset"
+            subj = "Your password has been changed"
+            body = f"Dear user, your password has been changed. Didn't do this? \
+                Secure your account by going to {link}."
+            send_email(current_identity.email, subj, body)
+            return ({"message": "Successfully changed password."}, 200)
+
+
+class RequestPasswordResetResource(Resource):
+    schema = RequestPasswordResetSchema()
+    
+    def post(self):
+        try:
+            res = self.schema.load(request.get_json())
+        except ValidationError as error:
+            field = list(error.messages.keys())[0]
+            return ({"message": f"{field}: {error.messages[field][0]}"}, 400)
+        
+        if not current_app.config["EMAIL_API_KEY"]:
+            return ({"message": "Could not send e-mail; API key not available."}, 500)
+
+        user = User.find_by_email(res["email"])
+        if not user:
+            # We don't want malicious users to know whether or not an e-mail is in our database.
+            return ({"message": "Recovery e-mail successfully sent."}, 200)
+        
+        token = user.get_password_reset_token()
+        link = f"https://app.gridt.org/reset_password?token={token}"
+        subj = "Reset your password"
+        body = f"Your password has been reset, please follow the following link: {link}"
+        send_email(user.email, subj, body)
+        return ({"message": "Recovery e-mail successfully sent."}, 200)
+
+
+class ResetPasswordResource(Resource):
+    schema = ResetPasswordSchema()
+    
+    def post(self):
+        secret_key = current_app.config["SECRET_KEY"]
+        try:
+            res = self.schema.load(request.get_json())
+        except ValidationError as error:
+            field = list(error.messages.keys())[0]
+            return ({"message": f"{field}: {error.messages[field][0]}"}, 400)
+
+        try:
+            token_decoded = jwt.decode(res["token"], secret_key, algorithms=["HS256"])
+        except jwt.ExpiredSignatureError:
+            msg = "Signature has expired."
+            return ({"message": msg}, 400)
+        except jwt.InvalidTokenError:
+            msg = "Invalid token."
+            return ({"message": msg}, 400)
```

Comment:
It is actually much more logical to put this in the schema, as it has to do with the direct validation of the input. It could be dealt with there.